group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java

```diff
@@ -64,11 +64,11 @@ public static class TargetAssignmentResult {
         /**
          * The new target assignment for the group.
          */
-        private final Map<String, Assignment> targetAssignment;
+        private final Map<String, MemberAssignment> targetAssignment;
 
         TargetAssignmentResult(
             List<CoordinatorRecord> records,
-            Map<String, Assignment> targetAssignment
+            Map<String, MemberAssignment> targetAssignment
```

Comment:
Do we really need `Assignment`? It seems `Assignment` is almost same to `MemberAssignment` except for the "unmodified" wrap. Maybe we can remove `Assignment` by adding following changes to `MemberAssignment`

1. make `MemberAssignment#targetPartitions` return unmodified object. this is the replacement for `Assignment#targetPartitions`
2. add setter to `MemberAssignment` for other use cases which try to update inner `targetPartitions`

WDYT?

Reply:
`MemberAssignment` is part of the public interface so I don't want to change it and I'd like to keep it immutable. For the context, `Assignment` will have more fields in the future when we implement client side assignors. This is why we have it. Otherwise, I would have dropped it and directly used the `Map<Uuid, Set<Interger>>`.

Reply:
> and I'd like to keep it immutable. 

Pardon me, do you mean the variables of `MemberAssignment` is immutable? if so, there are use cases that modify the inner collection. For example:

https://github.com/apache/kafka/blob/a68a1cce824a8346509d5194e0e43a3cb36ba09a/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/AbstractUniformAssignmentBuilder.java#L46

https://github.com/apache/kafka/blob/a68a1cce824a8346509d5194e0e43a3cb36ba09a/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java#L180

https://github.com/apache/kafka/blob/a68a1cce824a8346509d5194e0e43a3cb36ba09a/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java#L201

Reply:
Yeah, that's right. I should not have said immutable. What I meant is that once a `MemberAssignment` is returned by the assignor, it should not change anymore. Adding a setter to it would loosen the contract in my opinion.

Reply:
Got it, thanks for sharing.